server/routes/getOffences.ts

```diff
@@ -31,24 +38,41 @@ const getOffences = async (models, req: Request, res: Response, next: NextFuncti
     startDate.setDate(startDate.getDate() - 30);
     startDate = startDate.toISOString(); // 2020-08-08T12:46:32.276Z FORMAT // 30 days ago date
   }
-
-  const where: any = {
-    '$ChainEventType.chain$': chain,
-    '$ChainEventType.event_name$': 'offence'
+  let where: any = {
+    chain_event_type_id: `${chain}-offences-offence`
   };
+
+  if (stash) {
+    const stashCheck = { [Op.and]: Sequelize.literal(`event_data->>'offenders' LIKE '%${stash}%'`) };
+    where = Object.assign(where, stashCheck);
+  }
+
   if (startDate && endDate) {
     where.created_at = {
       [Op.between]: [startDate, endDate]
     };
   }
   const offences = await models.ChainEvent.findAll({
-    where,
-    include: [
-      { model: models.ChainEventType }
-    ]
+    where
+  });
+
+  let validators: { [key: string]: { [block: string]: any } } = {};
+
+  offences.forEach((ofc) => {
+    const event_data: IEventData = ofc.dataValues;
+    validators = {};
+    event_data.offenders.forEach((offender) => {
+      let key = offender;
+      if (validators.hasOwnProperty(key)) {
+        validators[key][ofc.block_number.toString()] = event_data.kind;
```

Comment:
My only qualm with this is that ALL events from this DB are just `${chain}-offences-offence` so this is kinda pointless. If we're trying to populate a list or graph of offences that a validator has done, we shoulld probably just create a map from `validator => Vec<BlockNumber>` for all the block numbers a validator offended at.

Reply:
Hi @drewstone I get your point and yeah at first glance it looks like a waste of json to keep the data in a dictionary for a single validator. So I can definitely map the block numbers in a vector and return it. 
Just wanted to bring one thing in your notice before making the change. Currently this route can be called with or without a stash id. In case a stash id is provided, we only send data for a single provided validator(if exists) else the route returns all the validator offences. Your proposed approach works wonder for the first case but in the second one we are forced into sending a dictionary of validators each containing the array of blocknumbers. 
For the sake of keeping the return same response, I have used same dictionary structure even for a single validator(which yeah I guess doesnt make sense). 
One other motivation for this structure was the other routes like getSlashes, getRewards etc.. as they were also using the same response structure so I went with the flow! Please let me know how you want to proceed with it. Thanks

Reply:
Lets roll with this as is, and ensure that the UI works.